<a href="https://colab.research.google.com/github/Grupo242022/EyCD/blob/main/Entregable_2_Nico_Dany_parte_1_ejercicio_1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **DIPLOMATURA 2022**

# EXPLORACIÓN Y CURACIÓN DE DATOS

## Entregable # 2 Parte 1

## GRUPO Nº24

## INTEGRANTES:
   - Diana Fonnegra
   - Clarisa Manzone
   - Nico Rosales
   - Daniel Rubio

En esta notebook, vamos a cargar el conjunto de datos de [la compentencia Kaggle](https://www.kaggle.com/dansbecker/melbourne-housing-snapshot) sobre estimación de precios de ventas de propiedades en Melbourne, Australia.

Utilizaremos el conjunto de datos reducido producido por [DanB](https://www.kaggle.com/dansbecker). Hemos subido una copia a un servidor de la Universidad Nacional de Córdoba para facilitar su acceso remoto.

In [37]:
import matplotlib.pyplot as plt
import numpy
import pandas as pd

import seaborn
seaborn.set_context('talk')

from sqlalchemy import create_engine, text

In [38]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


## Ejercicio 1 SQL: 

1. Crear una base de datos en SQLite utilizando la libreria [SQLalchemy](https://stackoverflow.com/questions/2268050/execute-sql-from-file-in-sqlalchemy).
https://docs.sqlalchemy.org/en/14/core/engines.html#sqlite

2. Ingestar los datos provistos en 'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/melb_data.csv' en una tabla y el dataset generado en clase con datos de airbnb y sus precios por codigo postal en otra.

3. Implementar consultas en SQL que respondan con la siguiente información:

    - cantidad de registros totales por ciudad.
    - cantidad de registros totales por barrio y ciudad.

4. Combinar los datasets de ambas tablas ingestadas utilizando el comando JOIN de SQL  para obtener un resultado similar a lo realizado con Pandas en clase.  

## Resolución:

### Lectura del dataset con los datos de las viviendas de Melbourne, tomados de la compentencia Kaggle sobre estimación de precios de ventas de propiedades en Melbourne, Australia. Procedemos a su lectura y carga en el dataframe "df_melb".

#### *Nota:* El sitio https://sqliteviewer.flowsoft7.com/ ofrece un entorno GUI para cargar una base sqlite, visualizar datos y ejecutar consultas. Es sumamente util para insepeccionar la información almacenada y depurar senstencias sql.

In [39]:
url_melb = 'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/melb_data.csv'
df_melb = pd.read_csv(url_melb)
# df_melb.dtypes

### Lectura del dataset Airbnb. El dataframe "df_airbnb" continene la información con las columnas de interes vistas en clases.

In [40]:
interesting_cols = [
  'description', 'neighborhood_overview',
  'street', 'neighborhood', 'city', 'suburb', 'state', 'zipcode',
  'price', 'weekly_price', 'monthly_price',
  'latitude', 'longitude',
]
df_airbnb = pd.read_csv(
    'https://cs.famaf.unc.edu.ar/~mteruel/datasets/diplodatos/cleansed_listings_dec18.csv',
    usecols=interesting_cols)

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:2882: DtypeWarning: Columns (35) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


## 1.1 Creamos la base de datos "melboure_db" en SQLite utilizando la libreria SQLalchemy. La misma esta alocada en gdrive.

In [65]:
engine = create_engine("sqlite:////content/drive/MyDrive/carpeta_db/melbourne_db.sqlite3", echo=True) 
# echo=True setea en adelante la visualización del log de comandos y acciones durante la ejecunción de una sentencia sql

## 1.2 Creación de tablas e ingesta de datos

Ingestamos en la tabla "kaggle_data" los datos del correspondiente dataframe.

In [66]:
df_melb.to_sql('melb_data', con=engine, if_exists="replace")

2022-05-28 21:37:23,105 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("melb_data")
2022-05-28 21:37:23,107 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:23,116 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("melb_data")
2022-05-28 21:37:23,118 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:23,134 INFO sqlalchemy.engine.Engine SELECT name FROM sqlite_master WHERE type='table' ORDER BY name
2022-05-28 21:37:23,136 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:23,141 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("melb_data")
2022-05-28 21:37:23,145 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:23,157 INFO sqlalchemy.engine.Engine SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
2022-05-28 21:37:23,161 INFO sqlalchemy.engine.Engine [raw sql] ('melb_data',)
2022-05-28 21:37:23,169 INFO sqlalchemy.engine.Engine PRAGMA main.foreign_key_list("m

Ingestamos en la tabla "airbnb_data" los datos del correspondiente dataframe.

In [67]:
df_airbnb.to_sql('airbnb_data', con=engine, if_exists="replace")

2022-05-28 21:37:28,703 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("airbnb_data")
2022-05-28 21:37:28,704 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:28,714 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("airbnb_data")
2022-05-28 21:37:28,716 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:28,728 INFO sqlalchemy.engine.Engine SELECT name FROM sqlite_master WHERE type='table' ORDER BY name
2022-05-28 21:37:28,730 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:28,740 INFO sqlalchemy.engine.Engine PRAGMA main.table_info("airbnb_data")
2022-05-28 21:37:28,741 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 21:37:28,750 INFO sqlalchemy.engine.Engine SELECT sql FROM  (SELECT * FROM sqlite_master UNION ALL   SELECT * FROM sqlite_temp_master) WHERE name = ? AND type = 'table'
2022-05-28 21:37:28,751 INFO sqlalchemy.engine.Engine [raw sql] ('airbnb_data',)
2022-05-28 21:37:28,753 INFO sqlalchemy.engine.Engine PRAGMA main.foreign_key

Relevamos los campos de cada tabla a fin de idenfiticar su composición y los tipos de datos aplicados por Pandas al ingestar los datos en la base de datos

In [68]:
# Consultamos los campos de la tabla "melb_data"
# La siguiente sentencia, con el nombre de la tabla entre corchetes, devuelve \
# atributos de interes de la tabla tales como el nombre de los campos \
# y tipos asociados entre otros.

query_campos_melb="PRAGMA table_info([melb_data])"
with engine.connect() as con:
      rs = con.execute(query_campos_melb)
      print("Los campos de la tabla melb_data son:")
      for row in rs:
          print(row)
      print('\n\n')

2022-05-28 21:39:20,905 INFO sqlalchemy.engine.Engine PRAGMA table_info([melb_data])
2022-05-28 21:39:20,909 INFO sqlalchemy.engine.Engine [raw sql] ()
Los campos de la tabla melb_data son:
(0, 'index', 'BIGINT', 0, None, 0)
(1, 'Suburb', 'TEXT', 0, None, 0)
(2, 'Address', 'TEXT', 0, None, 0)
(3, 'Rooms', 'BIGINT', 0, None, 0)
(4, 'Type', 'TEXT', 0, None, 0)
(5, 'Price', 'FLOAT', 0, None, 0)
(6, 'Method', 'TEXT', 0, None, 0)
(7, 'SellerG', 'TEXT', 0, None, 0)
(8, 'Date', 'TEXT', 0, None, 0)
(9, 'Distance', 'FLOAT', 0, None, 0)
(10, 'Postcode', 'FLOAT', 0, None, 0)
(11, 'Bedroom2', 'FLOAT', 0, None, 0)
(12, 'Bathroom', 'FLOAT', 0, None, 0)
(13, 'Car', 'FLOAT', 0, None, 0)
(14, 'Landsize', 'FLOAT', 0, None, 0)
(15, 'BuildingArea', 'FLOAT', 0, None, 0)
(16, 'YearBuilt', 'FLOAT', 0, None, 0)
(17, 'CouncilArea', 'TEXT', 0, None, 0)
(18, 'Lattitude', 'FLOAT', 0, None, 0)
(19, 'Longtitude', 'FLOAT', 0, None, 0)
(20, 'Regionname', 'TEXT', 0, None, 0)
(21, 'Propertycount', 'FLOAT', 0, None, 0)


In [45]:
query_campos_airbnb="PRAGMA table_info([airbnb_data])"
with engine.connect() as con:
      rs = con.execute(query_campos_airbnb)
      print("Los campos de la tabla airbnb_data son:")
      for row in rs:
          print(row)
      print('\n\n')

Los campos de la tabla airbnb_data son:
(0, 'index', 'BIGINT', 0, None, 0)
(1, 'description', 'TEXT', 0, None, 0)
(2, 'neighborhood_overview', 'TEXT', 0, None, 0)
(3, 'street', 'TEXT', 0, None, 0)
(4, 'neighborhood', 'TEXT', 0, None, 0)
(5, 'city', 'TEXT', 0, None, 0)
(6, 'suburb', 'TEXT', 0, None, 0)
(7, 'state', 'TEXT', 0, None, 0)
(8, 'zipcode', 'TEXT', 0, None, 0)
(9, 'latitude', 'FLOAT', 0, None, 0)
(10, 'longitude', 'FLOAT', 0, None, 0)
(11, 'price', 'BIGINT', 0, None, 0)
(12, 'weekly_price', 'FLOAT', 0, None, 0)
(13, 'monthly_price', 'FLOAT', 0, None, 0)





## 1.3 Consultas

### 1.3.a Tabla melb_data. Cantidad de registros totales por barrio

In [59]:
# Cantidad de registros totales por barrio de la tabla melb_data:
query_a = "SELECT Suburb, COUNT(1) as Total_records_Suburb FROM \
melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC"

In [60]:
with engine.connect() as con:
      rs = con.execute(query_a)
      print(query_a)
      for row in rs:
          print(row)

SELECT Suburb, COUNT(1) as Total_records_Suburb FROM melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC
('Reservoir', 359)
('Richmond', 260)
('Bentleigh East', 249)
('Preston', 239)
('Brunswick', 222)
('Essendon', 220)
('South Yarra', 202)
('Glen Iris', 195)
('Hawthorn', 191)
('Coburg', 190)
('Northcote', 188)
('Brighton', 186)
('Kew', 177)
('Balwyn North', 171)
('Pascoe Vale', 171)
('Yarraville', 164)
('St Kilda', 162)
('Glenroy', 159)
('Port Melbourne', 153)
('Moonee Ponds', 149)
('Carnegie', 146)
('Bentleigh', 139)
('Thornbury', 139)
('Brighton East', 138)
('Newport', 137)
('Camberwell', 132)
('Footscray', 132)
('Elwood', 131)
('Ascot Vale', 130)
('Hampton', 130)
('Surrey Hills', 125)
('Prahran', 123)
('Keilor East', 121)
('Hawthorn East', 119)
('Kensington', 117)
('Malvern East', 117)
('Doncaster', 116)
('Sunshine', 111)
('Brunswick West', 110)
('Balwyn', 107)
('Williamstown', 107)
('West Footscray', 105)
('Fawkner', 102)
('Toorak', 100)
('Ivanhoe', 98)
('Maribyrnong', 96

### 1.3.b Tabla melb_data. Cantidad de registros totales por barrio y ciudad

In [55]:
# Cantidad de registros totales por barrio y ciudad
query_b = "SELECT Suburb, CouncilArea, COUNT(1) as Total_records_Suburb FROM \
melb_data GROUP BY Suburb, CouncilArea ORDER BY Total_records_Suburb DESC"

In [58]:
with engine.connect() as con:
      rs = con.execute(query_b)
      for row in rs:
          print(row)
 

('Reservoir', 'Darebin', 323)
('Richmond', 'Yarra', 251)
('Bentleigh East', 'Glen Eira', 234)
('Preston', 'Darebin', 228)
('Brunswick', 'Moreland', 208)
('Essendon', 'Moonee Valley', 208)
('Hawthorn', 'Boroondara', 186)
('Brighton', 'Bayside', 176)
('Coburg', 'Moreland', 176)
('Northcote', 'Darebin', 175)
('Pascoe Vale', 'Moreland', 161)
('St Kilda', 'Port Phillip', 160)
('Kew', 'Boroondara', 157)
('South Yarra', 'Stonnington', 157)
('Yarraville', 'Maribyrnong', 154)
('Balwyn North', 'Boroondara', 151)
('Glenroy', 'Moreland', 144)
('Moonee Ponds', 'Moonee Valley', 139)
('Carnegie', 'Glen Eira', 138)
('Thornbury', 'Darebin', 132)
('Bentleigh', 'Glen Eira', 131)
('Newport', 'Hobsons Bay', 129)
('Footscray', 'Maribyrnong', 126)
('Ascot Vale', 'Moonee Valley', 125)
('Elwood', 'Port Phillip', 125)
('Brighton East', 'Bayside', 124)
('Camberwell', 'Boroondara', 124)
('Glen Iris', 'Boroondara', 117)
('Port Melbourne', 'Port Phillip', 117)
('Hampton', 'Bayside', 115)
('Hawthorn East', 'Boroonda

## 1.4 Combinación de los datasets de ambas tablas ingestadas utilizando el comando JOIN de SQL.

In [84]:
# Creamos el "campo zipcode_int" para alocar el codigo postal en formato entero
engine.execute("ALTER TABLE airbnb_data ADD COLUMN zipcode_int 'integer'")

2022-05-28 23:03:02,226 INFO sqlalchemy.engine.Engine ALTER TABLE airbnb_data ADD COLUMN zipcode_int 'integer'
2022-05-28 23:03:02,230 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 23:03:02,248 INFO sqlalchemy.engine.Engine COMMIT


In [87]:
# Ingestamos "zipcode_int" mediante CAST a entero del campo "zipcode"  
engine.execute("UPDATE airbnb_data SET zipcode_int = CAST(zipcode AS INTEGER)")

2022-05-28 23:10:30,835 INFO sqlalchemy.engine.Engine UPDATE airbnb_data SET zipcode_int = CAST(zipcode AS INTEGER)
2022-05-28 23:10:30,840 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-28 23:10:40,641 INFO sqlalchemy.engine.Engine COMMIT


In [90]:
# Relevamiento e idenficación de codigos postales
query_c = "SELECT zipcode_int, COUNT(zipcode_int) as cantidad FROM airbnb_data GROUP BY zipcode_int ORDER BY cantidad desc"

with engine.connect() as con:
      rs = con.execute(query_c)
      for row in rs:
          print(row)
 


2022-05-28 23:15:20,624 INFO sqlalchemy.engine.Engine SELECT zipcode_int, COUNT(zipcode_int) as cantidad FROM airbnb_data GROUP BY zipcode_int ORDER BY cantidad desc
2022-05-28 23:15:20,630 INFO sqlalchemy.engine.Engine [raw sql] ()
(3000, 3368)
(3006, 1268)
(3182, 1135)
(3141, 881)
(3004, 728)
(3121, 628)
(3008, 616)
(3053, 538)
(3181, 460)
(3056, 446)
(3065, 387)
(3051, 338)
(3184, 335)
(3183, 326)
(3068, 308)
(3205, 300)
(3066, 297)
(3070, 284)
(3003, 267)
(3030, 258)
(3067, 258)
(3207, 257)
(3128, 214)
(3011, 198)
(3002, 197)
(3150, 189)
(3031, 188)
(3777, 187)
(3057, 185)
(3122, 184)
(3032, 176)
(3163, 156)
(3168, 152)
(3072, 141)
(3108, 141)
(3054, 138)
(3058, 133)
(3186, 124)
(3161, 122)
(3206, 122)
(3142, 118)
(3145, 116)
(3071, 114)
(3029, 113)
(3101, 107)
(3185, 106)
(3129, 104)
(3199, 103)
(3052, 90)
(3788, 89)
(3799, 89)
(3124, 88)
(3055, 87)
(3083, 87)
(3013, 82)
(3123, 81)
(3162, 81)
(3204, 80)
(3149, 77)
(3012, 76)
(3016, 74)
(3078, 74)
(3084, 73)
(3146, 72)
(3188, 71)
(

Probamos las consultas


In [ ]:
with engine.connect() as con:
      rs = con.execute(query2)
      print(query2)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-26 20:54:24,384 INFO sqlalchemy.engine.Engine SELECT city as City, COUNT(1) as Total_records FROM melb_data_airbnb GROUP BY city ORDER BY Total_records DESC
2022-05-26 20:54:24,388 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT city as City, COUNT(1) as Total_records FROM melb_data_airbnb GROUP BY city ORDER BY Total_records DESC
('Melbourne', 7368)
('Port Phillip', 2808)
('Yarra', 2049)
('Stonnington', 1621)
('Moreland', 967)
('Yarra Ranges', 771)
('Darebin', 698)
('Boroondara', 664)
('Glen Eira', 631)
('Whitehorse', 614)
('Monash', 571)
('Maribyrnong', 436)
('Wyndham', 426)
('Bayside', 375)
('Moonee Valley', 344)
('Manningham', 313)
('Kingston', 309)
('Hobsons Bay', 239)
('Banyule', 203)
('Frankston', 177)
('Knox', 175)
('Hume', 170)
('Casey', 153)
('Greater Dandenong', 147)
('Whittlesea', 137)
('Cardinia', 123)
('Maroondah', 115)
('Brimbank', 108)
('Melton', 95)
('Nillumbik', 88)





Controlamos que funcione el código comparando la misma búsqueda con el df "df_airbnb"

In [ ]:
df_airbnb.city.value_counts()

Melbourne            7368
Port Phillip         2808
Yarra                2049
Stonnington          1621
Moreland              967
Yarra Ranges          771
Darebin               698
Boroondara            664
Glen Eira             631
Whitehorse            614
Monash                571
Maribyrnong           436
Wyndham               426
Bayside               375
Moonee Valley         344
Manningham            313
Kingston              309
Hobsons Bay           239
Banyule               203
Frankston             177
Knox                  175
Hume                  170
Casey                 153
Greater Dandenong     147
Whittlesea            137
Cardinia              123
Maroondah             115
Brimbank              108
Melton                 95
Nillumbik              88
Name: city, dtype: int64

Como funciona, ejecutamos las dos queries que faltan.

In [ ]:
with engine.connect() as con:
      rs = con.execute(query3)
      print(query3)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-26 20:55:01,698 INFO sqlalchemy.engine.Engine SELECT suburb as Suburb, COUNT(1) as Total_records_Suburb_airbnb FROM melb_data_airbnb GROUP BY Suburb ORDER BY Total_records_Suburb_airbnb DESC
2022-05-26 20:55:01,701 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT suburb as Suburb, COUNT(1) as Total_records_Suburb_airbnb FROM melb_data_airbnb GROUP BY Suburb ORDER BY Total_records_Suburb_airbnb DESC
('Melbourne', 4274)
('Southbank', 1236)
('South Yarra', 883)
('Saint Kilda', 805)
('Docklands', 612)
('Richmond', 570)
('Carlton', 537)
('Brunswick', 439)
('Fitzroy', 378)
('North Melbourne', 340)
('Elwood', 328)
('Collingwood', 295)
('South Melbourne', 288)
('Northcote', 283)
('West Melbourne', 263)
('Prahran', 262)
('Abbotsford', 258)
('Port Melbourne', 258)
('Fitzroy North', 238)
('St Kilda', 197)
('Windsor', 195)
('East Melbourne', 193)
('Point Cook', 191)
('Box Hill', 182)
('Hawthorn', 182)
('Saint Kilda East', 182)
('Brunswick East', 179)
('Glen Waverley', 174)
('Footscray', 1

In [ ]:
with engine.connect() as con:
      rs = con.execute(query1)
      print(query1)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-26 20:55:19,512 INFO sqlalchemy.engine.Engine SELECT Suburb, COUNT(1) as Total_records_Suburb FROM melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC
2022-05-26 20:55:19,516 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT Suburb, COUNT(1) as Total_records_Suburb FROM melb_data GROUP BY Suburb ORDER BY Total_records_Suburb DESC
('Reservoir', 359)
('Richmond', 260)
('Bentleigh East', 249)
('Preston', 239)
('Brunswick', 222)
('Essendon', 220)
('South Yarra', 202)
('Glen Iris', 195)
('Hawthorn', 191)
('Coburg', 190)
('Northcote', 188)
('Brighton', 186)
('Kew', 177)
('Balwyn North', 171)
('Pascoe Vale', 171)
('Yarraville', 164)
('St Kilda', 162)
('Glenroy', 159)
('Port Melbourne', 153)
('Moonee Ponds', 149)
('Carnegie', 146)
('Bentleigh', 139)
('Thornbury', 139)
('Brighton East', 138)
('Newport', 137)
('Camberwell', 132)
('Footscray', 132)
('Elwood', 131)
('Ascot Vale', 130)
('Hampton', 130)
('Surrey Hills', 125)
('Prahran', 123)
('Keilor East', 121)
('Hawthorn East', 11

Vamos a ver el contenido de "zipcode" de la tabla melb_data_airbnb para corroborar que coincidan con los valores de "Postcode" de melb_data con la que haremos el JOIN

In [ ]:
query4 = "SELECT zipcode, COUNT(1) as Records_of_zipcodes FROM melb_data_airbnb GROUP BY zipcode ORDER BY Records_of_zipcodes DESC "
with engine.connect() as con:
      rs = con.execute(query4)
      print(query4)
      for row in rs:
          print(row)

      print('\n\n')

2022-05-24 04:13:26,831 INFO sqlalchemy.engine.Engine SELECT zipcode, COUNT(1) as Records_of_zipcodes FROM melb_data_airbnb GROUP BY zipcode ORDER BY Records_of_zipcodes DESC 
2022-05-24 04:13:26,834 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT zipcode, COUNT(1) as Records_of_zipcodes FROM melb_data_airbnb GROUP BY zipcode ORDER BY Records_of_zipcodes DESC 
('3000.0', 2491)
('3000', 876)
('3006.0', 860)
('3182.0', 579)
('3182', 556)
('3004.0', 555)
('3141.0', 526)
('3008.0', 506)
('3006', 408)
('3121.0', 371)
('3141', 355)
('3053.0', 349)
('3181.0', 275)
('3056.0', 257)
('3121', 257)
('3065', 220)
('3003.0', 216)
('3051.0', 198)
('3030.0', 190)
('3053', 189)
('3056', 189)
('3205.0', 186)
('3181', 185)
('3183.0', 181)
('3184', 181)
('3004', 173)
('3128.0', 169)
('3065.0', 167)
('3066.0', 166)
('3070.0', 164)
('3068', 160)
('3067.0', 157)
('3184.0', 154)
('3068.0', 148)
('3207.0', 146)
('3183', 145)
(None, 142)
('3150.0', 141)
('3051', 140)
('3066', 131)
('3011.0', 128)
('3108.0', 1

Se observan que los dos resultados con mayor frecuencia son 3000.0 con 2491 apariciones y 3000 con 876, siendo en realidad el mismo código postal.<br>
Además, filtramos por zipcode y calculamos el promedio de los precios diarios, semanales y mensuales.

In [ ]:
query5 = """
SELECT zipcode,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
      )
WHERE zipcode IS NOT NULL
GROUP BY zipcode
ORDER BY zipcode DESC
"""
with engine.connect() as con:
      rs = con.execute(query5)
      print(query5)
      for row in rs:
          print(row)

2022-05-24 04:13:38,039 INFO sqlalchemy.engine.Engine 
SELECT zipcode,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
      )
WHERE zipcode IS NOT NULL
GROUP BY zipcode
ORDER BY zipcode DESC

2022-05-24 04:13:38,042 INFO sqlalchemy.engine.Engine [raw sql] ()

SELECT zipcode,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
      )
WHERE zipcode IS NOT NULL
GROUP BY zipcode
ORDER BY zipcode DESC

(30122.0, 75.0, 525.0, 2281.0)
(9540.0, 245.0, None, None)
(9004.0, 226.0, None, None)
(8938.0, 296.0, None, None)
(7367.0, 230.0, None, None)
(7266.0, 220.0

Ahora aplicamos esta consulta pero a la unión de las dos tablas

In [ ]:
query6 = """
SELECT * FROM melb_data
LEFT JOIN (
  SELECT zipcode ,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
  FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
        )
  WHERE zipcode IS NOT NULL
  GROUP BY zipcode
  ORDER BY zipcode DESC
          )
      ON melb_data_airbnb.zipcode = melb_data.Postcode
"""
with engine.connect() as con:
      rs = con.execute(query6)
      print(query6)
      for row in rs:
          print(row)

2022-05-26 21:16:24,786 INFO sqlalchemy.engine.Engine 
SELECT * FROM melb_data
LEFT JOIN (
  SELECT zipcode ,  
    AVG(price) as price_mean, 
    AVG(weekly_price) as weekly_price_mean, 
    AVG(monthly_price) as monthly_price_mean
  FROM (
    SELECT 
        CAST(zipcode AS float) as zipcode,
        price,
        weekly_price,
        monthly_price
    FROM melb_data_airbnb
        )
  WHERE zipcode IS NOT NULL
  GROUP BY zipcode
  ORDER BY zipcode DESC
          )
      ON melb_data_airbnb.zipcode = melb_data.Postcode

2022-05-26 21:16:24,790 INFO sqlalchemy.engine.Engine [raw sql] ()
2022-05-26 21:16:24,796 INFO sqlalchemy.engine.Engine ROLLBACK


OperationalError: ignored

In [ ]:
df_airbnb.columns.to_list()

['description',
 'neighborhood_overview',
 'street',
 'neighborhood',
 'city',
 'suburb',
 'state',
 'zipcode',
 'latitude',
 'longitude',
 'price',
 'weekly_price',
 'monthly_price']

In [ ]:
query= "SELECT zipcode FROM melb_data_airbnb"
with engine.connect() as con:
      rs = con.execute(query)
      print(query)
      for row in rs:
          print(row)
          break

2022-05-26 21:21:34,286 INFO sqlalchemy.engine.Engine SELECT zipcode FROM melb_data_airbnb
2022-05-26 21:21:34,288 INFO sqlalchemy.engine.Engine [raw sql] ()
SELECT zipcode FROM melb_data_airbnb
('3105',)


Este último punto no se puede visualizar por el OperationalError: RIGHT and FULL OUTER JOINs are not currently supported.